src/cfi.rs

```diff
@@ -89,6 +90,149 @@ impl<R: Reader> From<R> for DebugFrame<R> {
+#[derive(Clone, Debug)]
+pub struct ParsedEhFrameHdr<R: Reader> {
+    addr_size: u8,
+    section: R,
+
+    eh_frame_ptr: Pointer,
+    fde_count: u64,
+    table_enc: DwEhPe,
+    table: R,
+}
+
+impl<'input, Endian> EhFrameHdr<EndianBuf<'input, Endian>>
+where
+    Endian: Endianity,
+{
+    /// Constructs a new `EhFrameHdr` instance from the data in the `.eh_frame_hdr` section.
+    pub fn new(section: &'input [u8], endian: Endian) -> Self {
+        Self::from(EndianBuf::new(section, endian))
+    }
+}
+
+impl<R: Reader> EhFrameHdr<R> {
+    /// Parses this `EhFrameHdr` to a `ParsedEhFrameHdr`.
+    pub fn parse(&self, bases: &BaseAddresses, addr_size: u8) -> Result<ParsedEhFrameHdr<R>> {
+        let mut reader = self.0.clone();
+        let version = reader.read_u8()?;
+        assert_eq!(version, 1); // FIXME return error
+
+        let eh_frame_ptr_enc = DwEhPe(reader.read_u8()?);
+        let fde_count_enc = DwEhPe(reader.read_u8()?);
+        let table_enc = DwEhPe(reader.read_u8()?);
+        let eh_frame_ptr = parse_encoded_pointer(eh_frame_ptr_enc, bases, addr_size, &self.0, &mut reader)?;
+        let fde_count = parse_encoded_pointer(fde_count_enc, bases, addr_size, &self.0, &mut reader)?;
+        let fde_count = match fde_count {
+            Pointer::Direct(c) => c,
+            Pointer::Indirect(_) => unreachable!(),
+        };
+
+        Ok(ParsedEhFrameHdr {
+            addr_size,
+            section: self.0.clone(),
+
+            eh_frame_ptr,
+            fde_count,
+            table_enc,
+            table: reader,
+        })
+    }
+}
+
+impl<R: Reader> Section<R> for EhFrameHdr<R> {
+    fn section_name() -> &'static str {
+        ".eh_frame_hdr"
+    }
+}
+
+impl<R: Reader> From<R> for EhFrameHdr<R> {
+    fn from(section: R) -> Self {
+        EhFrameHdr(section)
+    }
+}
+
+impl<R: Reader> ParsedEhFrameHdr<R> {
+    /// Returns the address of the binary's `.eh_frame` section.
+    pub fn eh_frame_ptr(&self) -> Pointer {
+        self.eh_frame_ptr
+    }
+
+    /// Retrieves the CFI binary search table, if there is one.
+    pub fn table(&self) -> Option<EhHdrTable<R>> {
+        if self.table_enc == constants::DW_EH_PE_omit {
+            None
+        } else {
+            Some(EhHdrTable { hdr: self })
+        }
+    }
+}
+
+/// The CFI binary search table that is an optional part of the `.eh_frame_hdr` section.
+#[derive(Debug, Clone)]
+pub struct EhHdrTable<'a, R: Reader + 'a> {
+    hdr: &'a ParsedEhFrameHdr<R>,
+}
+
+impl<'a, R: Reader + 'a> EhHdrTable<'a, R> {
+    /// Returns a pointer to the FDE for the given address.
+    ///
+    /// This performs a binary search.
+    pub fn lookup(&self, address: u64, bases: &BaseAddresses) -> Result<Pointer> {
+        let size = match self.hdr.table_enc.format() {
+            constants::DW_EH_PE_uleb128 | constants::DW_EH_PE_sleb128
+                => return Err(Error::VariableLengthSearchTable),
```

Comment:
I don't see anything in https://refspecs.linuxfoundation.org/LSB_1.3.0/gLSB/gLSB/ehframehdr.html that suggests that using LEBs is not valid here, although it does break binary search, as you point out. How about a comment to the effect that it doesn't make sense because of the binary search stuff, but that it isn't strictly prohibited.

Reply:
Yeah, another tricky problem. Just like with `fde_count`, the compiler *could* in theory do something completely nonsensical here for no reason at all.

Again, I decided not to bother as that's what other implementations do ([llvm](https://github.com/llvm-mirror/libunwind/blob/6a5c438e9ef71c86d1d5717519c618dcf5e8100d/src/EHHeaderParser.hpp#L150), [libunwind (lol)](https://github.com/pathscale/libunwind/blob/vanilla_pathscale/src/dwarf/Gfind_unwind_table.c#L153)).

So I would say it's de facto prohibited.

Reply:
Makes sense. We should leave a comment to this effect here and ... *queue broken record* ... have test coverage for this case.

> libunwind (lol)

Wow...